sklearn/metrics/_classification.py

```diff
@@ -337,62 +337,76 @@ def confusion_matrix(
             raise ValueError("At least one label specified must be in y_true")
 
     if sample_weight is None:
-        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
+        sample_weight = xp.ones(y_true.shape[0], dtype=xp.int64, device=device_)
     else:
-        sample_weight = np.asarray(sample_weight)
+        sample_weight = xp.asarray(sample_weight, device=device_)
 
     check_consistent_length(y_true, y_pred, sample_weight)
 
-    n_labels = labels.size
+    # TODO: remove condition when torch supports the size attribute
+    if xp.__name__ == "array_api_compat.torch":
+        n_labels = xp.size(labels)
+    else:
+        n_labels = labels.size
     # If labels are not consecutive integers starting from zero, then
     # y_true and y_pred must be converted into index form
     need_index_conversion = not (
-        labels.dtype.kind in {"i", "u", "b"}
-        and np.all(labels == np.arange(n_labels))
-        and y_true.min() >= 0
-        and y_pred.min() >= 0
+        xp.isdtype(labels.dtype, ("signed integer", "unsigned integer", "bool"))
+        and xp.all(labels == xp.arange(n_labels, device=device_))
+        and xp.min(y_true) >= 0
+        and xp.min(y_pred) >= 0
     )
     if need_index_conversion:
         label_to_ind = {y: x for x, y in enumerate(labels)}
-        y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-        y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+        y_pred = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_pred], device=device_
+        )
+        y_true = xp.asarray(
+            [label_to_ind.get(x, n_labels + 1) for x in y_true], device=device_
+        )
 
     # intersect y_pred, y_true with labels, eliminate items not in labels
-    ind = np.logical_and(y_pred < n_labels, y_true < n_labels)
-    if not np.all(ind):
+    ind = xp.logical_and(y_pred < n_labels, y_true < n_labels)
+    if not xp.all(ind):
         y_pred = y_pred[ind]
         y_true = y_true[ind]
         # also eliminate weights of eliminated items
         sample_weight = sample_weight[ind]
 
     # Choose the accumulator dtype to always have high precision
-    if sample_weight.dtype.kind in {"i", "u", "b"}:
-        dtype = np.int64
+    if xp.isdtype(sample_weight.dtype, ("signed integer", "unsigned integer", "bool")):
+        dtype = xp.int64
     else:
-        dtype = np.float64
-
-    cm = coo_matrix(
-        (sample_weight, (y_true, y_pred)),
-        shape=(n_labels, n_labels),
-        dtype=dtype,
-    ).toarray()
+        dtype = xp.float64
+
+    if _is_numpy_namespace(xp):
+        cm = coo_matrix(
+            (sample_weight, (y_true, y_pred)),
+            shape=(n_labels, n_labels),
+            dtype=dtype,
+        ).toarray()
+    else:
+        cm = xp.zeros((n_labels, n_labels), dtype=dtype, device=device_)
+        for true, pred, weight in zip(y_true, y_pred, sample_weight):
+            cm[true, pred] += weight
```

Comment:
Is this performant enough? I think it could be, because we are mostly dealing with small matrices at this point. But finding another way might be better. I am not sure how to do this with the tools available in the spec though.

Reply:
I have the feeling that this loop will kill any computational benefit of array API support. We might as well ensure that `y_true` and `y_pred` are numpy arrays using `_convert_to_numpy` and rely on the `coo_matrix` trick instead. This would keep the code simpler.

That being said, I think convenience array API support for classification metrics that rely on confusion matrix internally is useful as discussed in https://github.com/scikit-learn/scikit-learn/pull/30439#issuecomment-2531072292.

Reply:
I agree that it seems like python loops do not go well with GPUs. However there doesn't seem to be an alternative with the array api because it doesn't support any sort of advanced indexing.

So either we might have to use the loop if we insist on following the array api or we could simply use the original code by utilizing the _convert_to_numpy as @ogrisel suggested.

Reply:
I ran this to check both options:

```python
import timeit

code_to_test = """
import torch
from sklearn.metrics import confusion_matrix
from sklearn.base import config_context

xp=torch

y_true = xp.randint(0, 2, (2**23,), dtype=xp.int32)
y_pred = xp.randint(0, 2, (2**23,), dtype=xp.int32)
with config_context(array_api_dispatch=True):
    confusion_matrix(y_true,y_pred)
"""

execution_time = timeit.timeit(code_to_test, number=1)
print(execution_time)
```

The results are pretty clear, I am really surprised by the difference:

with _convert_to_numpy:
2.0403239529696293

with python loop: 
115.27976658102125

The code for the python loop is the current state of this PR.
The code for the _convert_to_numpy version is in an extra branch: [here](https://github.com/StefanieSenger/scikit-learn/blob/be0419634cdd285dbbd6880e8770089253291540/sklearn/metrics/_classification.py#L423)

So, I think this speaks for the `_convert_to_numpy` option?

Reply:
I have executed this in Colab, and I had to massively reduce the size of the arrays, because otherwise Colab would interrupt itself. This was now tested on size `2**10`:

with python loop and device='cpu': 2.3 s

with python loop and device='cuda': 7.9 s

with _convert_to_numpy and device='cpu': 2.9 s

with _convert_to_numpy and device='cuda': 2.7 s

I'm not really sure how to interprete these results, especially since the large difference between both cpu solutions is not there anymore ....

Reply:
I just repeated the experiment using Scaleway VM (Type RENDER-S). I was able to execute the 2**23 array size versions for the python loop and got this:

with python loop:
execution_time for device cpu: 175.60608643199998
execution_time for device cuda: 468.15440756399994

with convert_to_numpy
execution_time for device cpu: 2.8455443269999705
execution_time for device cuda: 0.3635161219999645

So it seems the python loop is really having a bad effect, especially on cuda.

But there is still that problem with the `number=1` repetitions, which makes our results object to chance. I will re-craft the test and report results.

Reply:
I have now re-defined the test code. This is the test I ran using Scaleway VM (Type RENDER-S):

```python
import timeit

code_to_test = """
import torch
from sklearn.metrics import confusion_matrix
from sklearn.base import config_context

xp=torch

y_true = xp.randint(low=0, high=10, size=(int(1e5),), dtype=xp.int32, device="cpu")
y_pred = xp.randint(low=0, high=10, size=(int(1e5),), dtype=xp.int32, device="cpu")
with config_context(array_api_dispatch=True):
    confusion_matrix(y_true,y_pred)
"""

execution_time = timeit.timeit(code_to_test, number=50)
print(f'execution_time for device cpu: {execution_time}')

code_to_test = """
import torch
from sklearn.metrics import confusion_matrix
from sklearn.base import config_context

xp=torch

y_true = xp.randint(low=0, high=10, size=(int(1e5),), dtype=xp.int32, device="cuda")
y_pred = xp.randint(low=0, high=10, size=(int(1e5),), dtype=xp.int32, device="cuda")
with config_context(array_api_dispatch=True):
    confusion_matrix(y_true,y_pred)
"""

execution_time = timeit.timeit(code_to_test, number=50)
print(f'execution_time for device cuda: {execution_time}')
```

These are the results:

with python loop (current branch)
```
execution_time for device cpu: 69.83575903099972
execution_time for device cuda: 276.34000336899953

```

with convert_to_numpy ([this branch](https://github.com/StefanieSenger/scikit-learn/blob/be0419634cdd285dbbd6880e8770089253291540/sklearn/metrics/_classification.py#L423))
```
execution_time for device cpu: 1.9843785450002542
execution_time for device cuda: 0.4149802029996863
```

This even more confirms that we should use `convert_to_numpy` and then the coo sparse matrix instead of the python loop.

I am happy with this test now. Did I misunderstand anything or did I forget to consider anything?

Sorry for the verbosity of my posts, I am just discovering both these topics: gpu arrays and performance testing, so my thoughts were not all straightforward all the time.

Reply:
Yes I think the experiments you performed confirm our initial hypothesis that using python loops will have a negative impact on the performance of any array api related code. 

Also it makes sense that cuda is more affected because that involves synchronization and communication between the gpu device and the cpu code in the case of using loops. 

Thank you for running the experiments.